### PR TITLE
feat. 844280 Add mechanism for SDK to pass M symbols to vsc-pq

### PR DESCRIPTION
**Pull request by albertli-msft**
Add mechanism for SDK to pass M symbols to vscode-powerquery

https://user-images.githubusercontent.com/69623692/178483838-fda7e257-00af-4ecb-bcf7-31a133fedcfe.mp4

### CODE REVIEW

File: server/src/library/libraryUtils.ts [hunk 1]
```diff
@@ -70,7 +92,24 @@ function getOrCreateLibraryDefinitions(
         definitionsByLocale.set(locale, mapped);
     }
 
-    return PQP.MapUtils.assertGet(definitionsByLocale, locale);
+    let result: Map<string, PQLS.Library.TLibraryDefinition> = PQP.MapUtils.assertGet(definitionsByLocale, locale);
+
+    // we got module library for this definition call
+    if (currentModuleLibraryJson.length) {
```
Comment (JordanBoltonMN) [inline server/src/library/libraryUtils.ts:98]: How often will this condition be true? If that's a regular occurrence I don't love how we're shallow copying the map each call.

File: server/src/library/moduleLibraries.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+export interface ModuleLibraryTrieNodeCache {
+    localizedLibrary?: PQLS.Library.ILibrary;
+}
+
+export interface AddOneTextDocumentToModuleLibraryResultType {
+    closestLibraryJson: LibraryJson;
+    cache: ModuleLibraryTrieNodeCache;
+}
+
+class ModuleLibraryTrieNode {
```
Comment (mattmasson) [inline server/src/library/moduleLibraries.ts:15]: Trie?

File: server/src/library/libraryUtils.ts [hunk 3]
```diff
@@ -70,7 +92,24 @@ function getOrCreateLibraryDefinitions(
...
+        // make sure we do immutable operation over here
+        const base: Map<string, PQLS.Library.TLibraryDefinition> = result;
+        result = new Map();
+
+        for (const [key, value] of base) {
```
Comment (JordanBoltonMN) [inline server/src/library/libraryUtils.ts:103]: nitpick: `base.entries()`

File: server/src/library/moduleLibraries.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    remove(): void {
+        if (this.parent) {
+            this.parent.children.delete(this.currentPath);
+        }
+    }
+
+    collectTextDocumentBeneath(visitorContext: { textDocuments: TextDocument[] }): void {
+        for (const [, theNode] of this.children) {
```
Comment (JordanBoltonMN) [inline server/src/library/moduleLibraries.ts:59]: nit: `this.children.values()`

**Review by JordanBoltonMN — CHANGES_REQUESTED** [review]
As discussed in the meeting, we probably want to go with a different approach. Instead of modifying the standard library we should have our external type resolver be aware of other files.

The existing type resolver for the standard library is `ILibrary.externalTypeResolver`. This is later piped to instances of ISymbolProvider (the base of all of our LS providers). We probably want to update this PR to instead focus on enhancing the externalTypeResolver. Make it attempt a lookup on both the default library and on a module level.

File: client/src/extension.ts [hunk 5]
```diff
@@ -7,12 +7,13 @@ import * as vscode from "vscode";
...
 import * as CommandFn from "./commands";
 import * as Subscriptions from "./subscriptions";
+import { LibraryJson, VscodePowerQueryApi } from "./vscode-powerquery.api";
 import { CommandConstant } from "./commandConstant";
 
 const commands: vscode.Disposable[] = [];
 let client: LC.LanguageClient;
 
-export async function activate(context: vscode.ExtensionContext): Promise<void> {
+export async function activate(context: vscode.ExtensionContext): Promise<VscodePowerQueryApi> {
```
[the remark this answers is not in the document]
Comment (albertli-msft) [reply client/src/extension.ts]: PowerQueryApi

File: server/src/library/moduleLibraries.ts (hunk 2 above)
Comment (albertli-msft) [reply server/src/library/moduleLibraries.ts]: renamed to ModuleLibraryTreeNode

File: server/src/library/moduleLibraries.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
+import * as PQLS from "@microsoft/powerquery-language-services";
```
[the remark this answers is not in the document]
Comment (albertli-msft) [reply server/src/library/moduleLibraries.ts:4]: ahhhh my bad

File: server/src/library/libraryUtils.ts (hunk 3 above)
Comment (albertli-msft) [reply server/src/library/libraryUtils.ts]: we replace this by merely shallow copying at level of externalTypeResolver and libraryDefinitions

File: server/src/library/moduleLibraries.ts (hunk 4 above)
Comment (albertli-msft) [reply server/src/library/moduleLibraries.ts:91]: sure

**Comment by albertli-msft** [comment]
> As discussed in the meeting, we probably want to go with a different approach. Instead of modifying the standard library we should have our external type resolver be aware of other files.
> 
> The existing type resolver for the standard library is `ILibrary.externalTypeResolver`. This is later piped to instances of ISymbolProvider (the base of all of our LS providers). We probably want to update this PR to instead focus on enhancing the externalTypeResolver. Make it attempt a lookup on both the default library and on a module level.

Understood, and I updated changes follow what we discussed I believed; please have a check 😋

File: server/src/library/libraryUtils.ts (hunk 1 above)
Comment (albertli-msft) [reply server/src/library/libraryUtils.ts]: true, we replaced this with a shallow copy of its parent, plus there is also another level of cache happening at `vscode-powerquery/server/src/server.tsL#393`, which could also shrink the times we hit with this shallow copy

**Review by JordanBoltonMN — APPROVED** [review]

File: server/src/server.ts [hunk 7]
```diff
@@ -10,18 +10,27 @@ import { TextDocument } from "vscode-languageserver-textdocument";
...
+import { LibraryDefinitionsGetter } from "./library/libraryTypeResolver";
+import { ModuleLibraryTreeNode } from "./library/moduleLibraries";
 
 interface SemanticTokenParams {
     readonly textDocumentUri: string;
     readonly cancellationToken: LS.CancellationToken;
 }
 
+interface ModuleLibraryUpdatedParams {
+    workspaceUriPath: string;
```
[the remark this answers is not in the document]
Comment (albertli-msft) [reply server/src/server.ts:25]: Yes, we love readonly very much😆

File: server/src/server.ts [hunk 8]
```diff
@@ -358,8 +380,21 @@ function createAnalysis(
     traceManager: PQP.Trace.TraceManager,
     cancellationToken: LS.CancellationToken | undefined,
 ): PQLS.Analysis {
-    const localizedLibrary: PQLS.Library.ILibrary = SettingsUtils.getLocalizedLibrary();
-    document.uri;
+    const externalLibraryDefinitionsGetters: LibraryDefinitionsGetter[] = [];
```
[the remark this answers is not in the document]
Comment (albertli-msft) [reply server/src/server.ts]: got it renamed into getLocalizedModuleLibraryFromTextDocument

**Pull request merged**